### ISSUE TITLE
[REQUEST] log and visualize key performance parameters

### ISSUE THREAD

**Issue by Giszmo** [cameri/nostream#144]
**Pledge**

I pledge to pay $100 if the following gets implemented in its entirety:

**Is your feature request related to a problem? Please describe.**

I'm tweaking parameters, updating the relay, observing genuine demand fluctuation and probably DOS attacks and only have CPU load, Disk I/O, Bandwidth and some other parameters that are logged for the whole machine to judge if anything is wrong. That's not good for taking informed decisions.

**Describe the solution you'd like**

- [ ] Log key performance indicators
- [ ] Expose them through a nice interface (serverUrl/stats for example) maybe using grafana or similar

KPIs I would love to see:

- [ ] concurrent websockets open
- [ ] concurrent queries watched
- [ ] websockets opened/closed
- [ ] time from connect to first EOSE
- [ ] time from [e:[<singleEventId>] query to EOSE
- [ ] events served
- [ ] Standard system load parameters: CPU, Load, Memory, Disk I/O, Disk Usage, Bandwidth

For some of these parameters, aggregate functions like median, 95th percentile ... would be of interest.

**Comment by bleetube** [comment]
I'll also add $100 (of sats) to this pledge.

**Comment by jmarbach** [comment]
@Giszmo @bleetube I've shared your request with the Grafana staff to raise awareness on building a metrics collector. No promises.

Meanwhile, if you just ned a metrics endpoint, you can forward whatever Prometheus metrics you'd like to long term storage in Grafana Cloud's managed Mimir service using the Influx proxy: 

https://grafana.com/docs/grafana-cloud/data-configuration/metrics/metrics-influxdb/push-from-telegraf/

**Comment by bleetube** [comment]
Actually I realized I can do the MVP part I described without touching nodejs.  I can just write some python to talk directly to postgres. I'll start working on it this weekend.

**Comment by kushagra0902** [comment]
> what are you thoughts on using OTEL instead for pushing metrics?

Yes will be a good architecture. Just a small follow-up, is our main intent data observability or real-time performance tracking, as I think a complete OTEL stack of collector and all the necessary modules will increase the overhead.
If it is just data observability, then I think we can go with a normal PSQL-based architecture

Please correct me if I am wrong at any point.

**Comment by cameri** [comment]
@kushagra0902  let's go with PSQL based architecture

**Comment by Ferryx349** [comment]
Hi, I believe this issue is now addressed. The logging and visualization work for nostream’s admin/observability stack is being implemented over the last few months.

### What #144 asked for and what shipped

**1. Log key performance indicators**  
Relay metrics are collected via OpenTelemetry and exposed to Prometheus, including:
- WebSocket connection counts
- Events accepted/rejected (rate and totals)
- Process CPU and memory per worker

**2. Expose KPIs through a usable interface**  
Operators can view live metrics in two ways:
- **Admin console** (`/admin/`) → Metrics tab with live SSE snapshots from `/admin/metrics`
- **Grafana** → `nostream-overview` dashboard (embedded in admin UI and available via docker-compose stack)

**3. KPIs from the original checklist**
| Requested KPI | Where it’s visible |
|---|---|
| Concurrent WebSockets | `Active connections`, `WebSocket connections (15m)` |
| Events served | `Events / sec`, `Accepted total` |
| System load (CPU/memory) | `CPU load %`, `Memory by worker`, `CPU by worker` |
| Rejection visibility | `Rejected / sec`, `Rejected total` |

### Implementation PRs

**Metrics foundation**
- #653 :- OpenTelemetry metrics bootstrap, Prometheus sidecar
- #661 :-  Instrument event, WebSocket handlers with OTEL metrics

**Admin API and live metrics**
- #641 — Admin backend foundation (auth, session, health)
- #662 — Prometheus-backed `/admin/metrics` SSE endpoint

**Visualization UI**
- #666 :- Observability dashboard UI with Grafana embeds

**Admin console extensions**
- #672 :- Shared settings-config module (CLI + admin UI)
- #690 :- Authenticated settings API endpoints
- #706 :- Settings editor tab in admin dashboard

### How to try it

1. Enable admin in settings and set `ADMIN_PASSWORD`
2. Run the observability stack (`docker compose up`) for Prometheus/Grafana
3. Open `/admin/` → Metrics tab for live KPIs, or use Grafana `nostream-overview`

---

If this covers the scope of #144, I’d appreciate closing this issue. Happy to adjust anything if something is still missing from the original checklist.

**Comment by cameri** [comment]
Thank you @Ferryx349 !
